server/core/terraform/terraform_client.go

```diff
@@ -376,100 +378,19 @@ type Line struct {
 // Callers can use the input channel to pass stdin input to the command.
 // If any error is passed on the out channel, there will be no
 // further output (so callers are free to exit).
-func (c *DefaultClient) RunCommandAsync(ctx command.ProjectContext, path string, args []string, customEnvVars map[string]string, v *version.Version, workspace string) (chan<- string, <-chan Line) {
-	outCh := make(chan Line)
-	inCh := make(chan string)
-
-	// We start a goroutine to do our work asynchronously and then immediately
-	// return our channels.
-	go func() {
-
-		// Ensure we close our channels when we exit.
-		defer func() {
-			close(outCh)
-			close(inCh)
-		}()
-
-		tfCmd, cmd, err := c.prepCmd(ctx.Log, v, workspace, path, args)
-		if err != nil {
-			ctx.Log.Err(err.Error())
-			outCh <- Line{Err: err}
-			return
-		}
-		stdout, _ := cmd.StdoutPipe()
-		stderr, _ := cmd.StderrPipe()
-		stdin, _ := cmd.StdinPipe()
-		envVars := cmd.Env
-		for key, val := range customEnvVars {
-			envVars = append(envVars, fmt.Sprintf("%s=%s", key, val))
-		}
-		cmd.Env = envVars
-
-		ctx.Log.Debug("starting %q in %q", tfCmd, path)
-		err = cmd.Start()
-		if err != nil {
-			err = errors.Wrapf(err, "running %q in %q", tfCmd, path)
-			ctx.Log.Err(err.Error())
-			outCh <- Line{Err: err}
-			return
-		}
-
-		// If we get anything on inCh, write it to stdin.
-		// This function will exit when inCh is closed which we do in our defer.
-		go func() {
-			for line := range inCh {
-				ctx.Log.Debug("writing %q to remote command's stdin", line)
-				_, err := io.WriteString(stdin, line)
-				if err != nil {
-					ctx.Log.Err(errors.Wrapf(err, "writing %q to process", line).Error())
-				}
-			}
-		}()
-
-		// Use a waitgroup to block until our stdout/err copying is complete.
-		wg := new(sync.WaitGroup)
-		wg.Add(2)
-		// Asynchronously copy from stdout/err to outCh.
-		go func() {
-			s := bufio.NewScanner(stdout)
-			buf := []byte{}
-			s.Buffer(buf, BufioScannerBufferSize)
-
-			for s.Scan() {
-				message := s.Text()
-				outCh <- Line{Line: message}
-				c.projectCmdOutputHandler.Send(ctx, message, false)
-			}
-			wg.Done()
-		}()
-		go func() {
-			s := bufio.NewScanner(stderr)
-			for s.Scan() {
-				message := s.Text()
-				outCh <- Line{Line: message}
-				c.projectCmdOutputHandler.Send(ctx, message, false)
-			}
-			wg.Done()
-		}()
-
-		// Wait for our copying to complete. This *must* be done before
-		// calling cmd.Wait(). (see https://github.com/golang/go/issues/19685)
-		wg.Wait()
-
-		// Wait for the command to complete.
-		err = cmd.Wait()
+func (c *DefaultClient) RunCommandAsync(ctx command.ProjectContext, path string, args []string, customEnvVars map[string]string, v *version.Version, workspace string) (chan<- string, <-chan models.Line, error) {
+	cmd, envVars, err := c.prepCmd(ctx.Log, v, workspace, path, args)
+	if err != nil {
+		return nil, nil, err
```

Comment:
An alternative to changing the signature of this `RunCommandAsync` method would be to create a dummy channel that we send an `%Line{Err: err}` message on. To be strictly correct we'd then also need to drain the `stdin` channel

I'm not sure that it's actually valuable to have a distinction from "unable to get the terraform version to run the command"  and "there was a problem running the terraform command".